File: src/main/java/com/gmail/borlandlp/minigamesdtools/geoip/ipstack/IpstackRequestBuilder.java
package com.gmail.borlandlp.minigamesdtools.geoip.ipstack;

import com.gmail.borlandlp.minigamesdtools.util.HttpUtils;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

public class IpstackRequestBuilder {
    private String ip;
    private String apiKey;

    public IpstackRequestBuilder setIp(String ip) {
        this.ip = ip;
        return this;
    }

    public IpstackRequestBuilder setApiKey(String apiKey) {
        this.apiKey = apiKey;
        return this;
    }

    public IpstackResponse doRequest() {
        String strResponse = null;
        try {
            strResponse = HttpUtils.doGetRequest("http://api.ipstack.com/" + ip + "?access_key=" + apiKey);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }

        JSONObject json = null;
        try {
            json = (JSONObject) (new JSONParser().parse(strResponse));
        } catch (ParseException e) {
            e.printStackTrace();
            return null;
        }

        IpstackResponse response = new IpstackResponse();
        if(json.containsKey("success")) {
            JSONObject obj = (JSONObject) json.get("error");
            int errCode = (int) (Long.parseLong(obj.get("code").toString())); // err codes range 0 - 500
            response.setError(Errors.getByErrCode(errCode));
        } else {
            response.setResponse(json);
        }

        return response;
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/geoip/ipstack/IpstackRequestBuilder.kt
package com.gmail.borlandlp.minigamesdtools.geoip.ipstack

import com.gmail.borlandlp.minigamesdtools.geoip.ipstack.ErrorResponse.Companion.getByErrCode
import com.gmail.borlandlp.minigamesdtools.util.HttpUtils
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import org.json.simple.parser.ParseException
import sun.net.ConnectionResetException

class IpstackRequestBuilder {
    private var ip: String? = null
    private var apiKey: String? = null

    fun setIp(ip: String?): IpstackRequestBuilder {
        this.ip = ip
        return this
    }

    fun setApiKey(apiKey: String): IpstackRequestBuilder {
        this.apiKey = apiKey
        return this
    }

    /**
     * @throws ConnectionResetException
     * @throws ParseException
     */
    fun doRequest(): IpstackResponse {
        var strResponse: String?
        val response = IpstackResponse()
        strResponse = try {
            HttpUtils.doGetRequest("http://api.ipstack.com/$ip?access_key=$apiKey")
        } catch (e: Exception) {
            e.printStackTrace()
            return response.apply {
                response.error = ErrorResponse.NETWORK_ERROR
            }
        }

        val json = try {
            JSONParser().parse(strResponse) as JSONObject
        } catch (exception: ParseException) {
            exception.printStackTrace()
            return response.apply {
                response.error = ErrorResponse.PARSE_ERROR
            }
        }

        if (json.containsKey("success")) {
            val obj = json["error"] as JSONObject?
            val errCode = obj!!["code"].toString().toLong().toInt() // err codes range 0 - 500
            response.error = getByErrCode(errCode)
        } else {
            response.setResponse(json)
        }

        return response
    }
}